### PR TITLE
[charts] Add `onClick` support

### LINKED ISSUE #10549: [charts] Pass `onClick` to chart item
Follow up to #10506

Allows to add `onClick` which should be triggered when clicking on an item

## Charts todo

- [x] Pie chart
- [ ] Line chart
- [ ] Bar chart
- [ ] scatter chart

## How

Follow #10506 to get inspiration. In short:

- The `onClick` goes the the chart component and the plot components
- it takes as an argument: the event, the item identifier, and the item content
- modify the cursor based on the presence or not of an `onClick`

**Search keywords**:

### PR DESCRIPTION
**Pull request by alexfauquette**
Fixes #10549

This PR is have commits per component impacted. I recommend to have first a look a documentation for the DX aspect, and a per commit review for code quality

- https://deploy-preview-11411--material-ui-x.netlify.app/x/react-charts/scatter/#click-event
- https://deploy-preview-11411--material-ui-x.netlify.app/x/react-charts/lines/#click-event
- https://deploy-preview-11411--material-ui-x.netlify.app/x/react-charts/bars/#click-event
- https://deploy-preview-11411--material-ui-x.netlify.app/x/react-charts/pie/#click-event

## Decision

I went with the `onAxisClick`/`onItemClick` such that easy usecase are handled with easy solution. The more complex ones are:
- scatter plot for which I extend the the voronoid handler
- line and bar chart that are the one using the `onAxisClick`

For docs, I'm moving to a docs per component to explain edge cases only in the docs of the charts for which they are relevant. It's easier to explain why `onClick` is more complex than usual in the scatter charts than in a dedicated page

---

It's a attempt to take best of the different proposed PRs

## Current solution

I'm adding an `<ChartsHandleOnClick onclick={onClick} />` that can be added anywhere.

This simply contain a `useEffect` that adds an event listener to the `svgRef` which will call the `onClick` callback with the event, plus some data if an item or an axis is active according to our internal state.

## Why such a solution

The `onClick` is not passed to each element, but listen at the `<svg/>` level for multiple reasons:

- For scatter charts, points are too small, so it's more interesting to benefit from the voronoid cells. And that can not be done when listening to event on components rendered by the scatter plot.
- For stacked charts, or charts with band/point axis, it's not clear if you want the data of the item you clicked, or all the items at a given axis position. See the recording of PR #10575 to get a good example about where axis data is more interesting than item ones.

## Issues

The main drawback I can see is the mouse pointer.

1. The event being at the `<svg/>` level to modify the mouse pointer when there is something to click I would need to modify style by looking at the context to see if there is an interactive item or axis selected.
2. Since the `onClick` returns multiple values, I can not know if devs will use only item data, or axis data. So I can't put a meaning full style.

## Ideas

I've some ideas of improvement

### Idea 1

The first one is to distinguish item and axis click with distinct `onItemClick` and `onAxisClick`

This should allow to setup good default styling:
- If `onAxisClick` is defined the drawing area get `cursor: pointer`
- If `onItemClick` is defined the drawing area get `cursor: pointer` when the active `item` is not `null`.

### Idea 2

A second one is to add `onClick` to element slots with additional data. Such that for simple use cases, Devs could do 

```jsx
<BarChart
	slotProps={{ 
		bar: { 
			onClick: (event, {dataIndex, seriesId, value}) => {...} 
		}
	}
/>
```

## Note

For now I did not update the Pie chart interaction, but will have to be coherent with other charts API

### Changelog

- 🚀 All charts have click handler. Test their respective docs demonstration to know more about the data format:

  - [scatter chart](https://next.mui.com/x/react-charts/scatter/#click-event)
  - [lines chart](https://next.mui.com/x/react-charts/lines/#click-event)
  - [bars chart](https://next.mui.com/x/react-charts/bars/#click-event)
  - [pie chart](https://next.mui.com/x/react-charts/pie/#click-event)

Big thanks to @giladappsforce and @yaredtsy for their contribution on exploring this feature 

### Breaking change

Pie charts `onClick` get renamed `onItemClick` for consistency with other charts click callback.

### CODE REVIEW

File: docs/data/charts/tooltip/tooltip.md [hunk 1]
```diff
@@ -53,6 +53,41 @@ This kind of interaction is controlled by series properties `highlightScope` whi
...
+The next demo allows to experiment the data received by `onClick` with various charts.
+
+Notice that `axis` is not defined for pie and scatter charts.
+You can also get `axis` and no `item` in bar and line charts by clicking outside of elements.
+
+{{"demo": "ElementClickNoSnap.js"}}
+
+#### Event management
+
+The `event` is the "onClick" event received by the `<svg/>` element.
```
Comment (flaviendelangle) [inline docs/data/charts/tooltip/tooltip.md:85]: The following is closer to the actual behavior right?

```suggestion
The `event` is the "onClick" event emitted by the `<svg/>` element.
```

But I think we could rework this section a bit to be clearer.

File: packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  item: null | ItemInteractionData<ChartSeriesType>,
+  series: FormattedSeries,
+): null | {
+  dataIndex?: number;
+  series: ChartSeriesDefaultized<ChartSeriesType>;
+} {
+  if (item === null) {
+    return null;
+  }
+  return { series: series[item.type]?.series[item.seriesId]!, dataIndex: item.dataIndex };
```
Comment (flaviendelangle) [inline packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx:44]: Are we sure about this `!`?
If we are not confident enough and we use `?` before, then the value can probably be undefined.

File: packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      onClick(event, { item: itemData, axis: { dataIndex, axisValue, seriesValues } });
+    };
+
+    element.addEventListener('click', handleMouseClick);
+    return () => {
+      element.removeEventListener('click', handleMouseClick);
+    };
+  }, [axis, item, series, svgRef, xAxisIds, yAxisIds, xAxis, yAxis, onClick]);
+
+  return <g />;
```
Comment (flaviendelangle) [inline packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx:100]: What is the goal of the `<g />` element?

Comment (alexfauquette) [reply packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx:100]: One script fails without. If I remember correctly it's proptypes that are not generated

Comment (flaviendelangle) [reply packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx:100]: If you return `null` `PropTypes` are not generated?

That's weird :grimacing:

Comment (alexfauquette) [reply packages/x-charts/src/ChartsOnClickHandler/ChartsOnClickHandler.tsx:100]: The empty fragment works :+1:

File: docs/data/charts/lines/lines.md [hunk 4]
```diff
@@ -90,6 +90,50 @@ However, it cannot extrapolate the curve before the first non-null data point or
...
+```
+
+{{"demo": "LineClickNoSnap.js"}}
+
+:::info
+Their is a slight difference between the `event` of `onAxisClick` and the others:
+
+- For `onAxisClick` it's a native mouse event emitted by the svg component.
+- For others, it's a React synthetic mouse event emitted by the area, line, or mark component.
+  :::
```
Comment (alexfauquette) [inline docs/data/charts/lines/lines.md:118]:
```suggestion

:::
```

File: docs/data/charts/bars/bars.md [hunk 5]
```diff
@@ -63,6 +63,47 @@ If you're using [composition](/x/react-charts/composition/), you should set the
...
+```
+
+{{"demo": "BarClickNoSnap.js"}}
+
+:::info
+Their is a slight difference between the `event` of `onItemClick` and `onAxisClick`:
+
+- For `onItemClick` it's a React synthetic mouse event emitted by the bar component.
+- For `onAxisClick` it's a native mouse event emitted by the svg component.
+  :::
```
Comment (alexfauquette) [inline docs/data/charts/bars/bars.md:89]:
```suggestion
- For `onAxisClick` it's a native mouse event emitted by the svg component.

:::
```

File: packages/x-charts/src/LineChart/LinePlot.tsx [hunk 6]
```diff
@@ -18,7 +19,17 @@ export interface LinePlotSlotProps {
 
 export interface LinePlotProps
   extends React.SVGAttributes<SVGSVGElement>,
-    Pick<LineElementProps, 'slots' | 'slotProps'> {}
+    Pick<LineElementProps, 'slots' | 'slotProps'> {
+  /**
+   * Callback fired when a pie item is clicked.
+   * @param {React.MouseEvent<SVGPathElement, MouseEvent>} event The event source of the callback.
+   * @param {LineItemIdentifier} lineItemIdentifier The pie item identifier.
```
Comment (LukasTy) [inline packages/x-charts/src/LineChart/LinePlot.tsx:26]:
```suggestion
   * @param {LineItemIdentifier} lineItemIdentifier The line item identifier.
```

File: packages/x-charts/src/LineChart/MarkPlot.tsx [hunk 7]
```diff
@@ -25,6 +26,15 @@ export interface MarkPlotProps extends React.SVGAttributes<SVGSVGElement> {
    * @default {}
    */
   slotProps?: MarkPlotSlotProps;
+  /**
+   * Callback fired when a pie item is clicked.
```
Comment (LukasTy) [inline packages/x-charts/src/LineChart/MarkPlot.tsx:30]:
```suggestion
   * Callback fired when a line mark item is clicked.
```

File: packages/x-charts/src/LineChart/MarkPlot.tsx [hunk 8]
```diff
@@ -25,6 +26,15 @@ export interface MarkPlotProps extends React.SVGAttributes<SVGSVGElement> {
...
+   * @param {React.MouseEvent<SVGPathElement, MouseEvent>} event The event source of the callback.
+   * @param {LineItemIdentifier} lineItemIdentifier The pie item identifier.
```
Comment (LukasTy) [inline packages/x-charts/src/LineChart/MarkPlot.tsx:32]:
```suggestion
   * @param {LineItemIdentifier} lineItemIdentifier The line item identifier.
```

File: packages/x-charts/src/ScatterChart/Scatter.tsx [hunk 9]
```diff
@@ -17,6 +17,15 @@ export interface ScatterProps {
   yScale: D3Scale;
   markerSize: number;
   color: string;
+  /**
+   * Callback fired when clicking on a scatter item.
+   * @param {MouseEvent} event Mouse event catched at the svg level
```
Comment (LukasTy) [inline packages/x-charts/src/ScatterChart/Scatter.tsx:22]:
```suggestion
   * @param {MouseEvent} event Mouse event recorded at the SVG level.
```

File: packages/x-charts/src/ScatterChart/Scatter.tsx [hunk 10]
```diff
@@ -17,6 +17,15 @@ export interface ScatterProps {
...
+   * @param {ScatterItemIdentifier} scatterItemIdentifier Identify whihc item got clicked
```
Comment (LukasTy) [inline packages/x-charts/src/ScatterChart/Scatter.tsx:23]: For consistency. 🤔 
```suggestion
   * @param {ScatterItemIdentifier} scatterItemIdentifier The scatter item identifier.
```

File: packages/x-charts/src/ScatterChart/ScatterChart.tsx [hunk 11]
```diff
@@ -64,6 +69,12 @@ export interface ScatterChartProps
    * @default {}
    */
   slotProps?: ScatterChartSlotProps;
+  /**
+   * Callback fired when clicking on a scatter item.
+   * @param {MouseEvent} event Mouse event catched at the svg level if using voronoid cells. Or Mouse event from the scatter element `disableVoronoi=true`.
+   * @param {ScatterItemIdentifier} scatterItemIdentifier Identify whihc item got clicked
```
Comment (LukasTy) [inline packages/x-charts/src/ScatterChart/ScatterChart.tsx:75]:
```suggestion
   * @param {ScatterItemIdentifier} scatterItemIdentifier The scatter item identifier.
```

File: packages/x-charts/src/ScatterChart/ScatterChart.tsx [hunk 12]
```diff
@@ -64,6 +69,12 @@ export interface ScatterChartProps
    * @default {}
    */
   slotProps?: ScatterChartSlotProps;
+  /**
+   * Callback fired when clicking on a scatter item.
+   * @param {MouseEvent} event Mouse event catched at the svg level if using voronoid cells. Or Mouse event from the scatter element `disableVoronoi=true`.
```
Comment (LukasTy) [inline packages/x-charts/src/ScatterChart/ScatterChart.tsx:74]:
```suggestion
   * @param {MouseEvent} event Mouse event recorded at the SVG level if using Voronoi cells. Or Mouse event from the scatter element, when `disableVoronoi=true`.
```

**Review by LukasTy — COMMENTED** [review]
Superb work! 👍 👏 😍 

Leaving some general nitpicks and raising a discussion point below.

Have you considered going with one event handler with a type prop to identify the type on the consumer side? 🤔 
I've checked and it seems that both Nivo and Highcharts have only a single `click` handler. 🤷 
At a glance it seems that we are adding quite a bit of complexity and it's a bit unclear if it is warranted. 🤔

**Comment by alexfauquette** [comment]
> Have you considered going with one event handler with a type prop to identify the type on the consumer side? 🤔

I made test for line chart:
- For Recharts: they only add it to the container and behave a similar way to the onAxisClick I'm adding
- For Nivo: The onClick works with voronoid but not with the vertical slices (you can only click on one point at the time)
- For highchart, I did not found the docs about how the manage click 🙈

The complexity added by having multiple way of listening click event allows to adapt the chart behavior. For example modifying the pointer on hover if something is listening to the click.

I think the complexity is the largest for those components. Tree view, gauge, and other don't have those axes triggering issues, so they will only get itemClick

**Comment by LukasTy** [comment]
Gotcha. Seems fair. 👌 

> I think the complexity is the largest for those components. Tree view, gauge, and other don't have those axes triggering issues, so they will only get itemClick

Do you feel that it still makes sense to go for consistent naming (`onItemClick` vs `onClick`) even though we know that there will only be one kind of click handler? 🤔

File: docs/data/charts/pie/PieClickNoSnap.js [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+          height={300}
+          slotProps={{
+            legend: { hidden: true },
+          }}
+          onItemClick={(event, d) => setItemData(d)}
+        />{' '}
+      </Box>
+
+      <Stack direction="column" sx={{ width: { xs: '100%', md: '40%' } }}>
+        <Typography>click data</Typography>
```
[the remark this answers is not in the document]
Comment (noraleonte) [reply docs/data/charts/pie/PieClickNoSnap.js:67]:
```suggestion
             <Box sx={{display:'flex', justifyContent:'space-between'}}>
        <Typography>Click on the chart!</Typography>
        <IconButton aria-label="reset">  // reset button
          <UndoOutlinedIcon/>
        </IconButton>
      </Box>
```
I agree, the reset button would make the experience a lot better 👌 We can do something simple, like this. And some slight improvement of the text

File: docs/data/charts/pie/PieClickNoSnap.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        <HighlightedCode
+          code={`// Data from item click
+${itemData ? JSON.stringify(itemData, null, 2) : '// click on the chart'}
+`}
```
Comment (noraleonte) [inline docs/data/charts/pie/PieClickNoSnap.js:71]:
```suggestion
          code={
itemData ? `// Data from item click:
${JSON.stringify(itemData, null, 2)}` : '// The data will appear here'
}
```
Maybe rephrasing these comments a bit could make the expected outcome a bit clearer 🤔

**Review by noraleonte — COMMENTED** [review]
Amazing work! 🍰 🎉 

Leaving a small suggestion to follow up on Lukas' comment 👍

**Comment by alexfauquette** [comment]
> Do you feel that it still makes sense to go for consistent naming (onItemClick vs onClick) even though we know that there will only be one kind of click handler? 🤔

It feel safer because we could have `<PieChart />` getting `onLegendClick`, end other event handler

**Comment by wascou** [comment]
Hey thanks for the work, but one important thing is missing :)
Composability!
May you extend this to a nested _ScatterPlot_ in a _ResponsiveChartContainer_?

**Comment by alexfauquette** [comment]
Thanks @wascou  for the heads up, I forgot to document how to use composition with scatter charts. The best documentation for now might be the code itself :)

https://github.com/mui/mui-x/blob/00b6b27380611d81e4c0ae16fc117047e2b70594/packages/x-charts/src/ScatterChart/ScatterChart.tsx#L127-L130

https://github.com/mui/mui-x/blob/00b6b27380611d81e4c0ae16fc117047e2b70594/packages/x-charts/src/ScatterChart/ScatterChart.tsx#L141-L145

The `<ResponsiveChartContainer />` shoudl have no impact on the click. And scatter plot has the same behavior as other. But I would recommend to use the click listener of the `<ChartsVoronoiHandler />`

Both are available only in v7

**Comment by wascou** [comment]
ah ok, no BC for this one ? I have to wait for a stable v7 then

**Comment by wascou** [comment]
Ok, works perfectly on my side. Voronoi i interresting but not very accurate when you have a lot of points in your scatter plot. I also put the `onItemClick` on ScatterPlot. Well done @alexfauquette and @LukasTy !

**Pull request merged**